### ISSUE TITLE
CDB topology rewrite + remaining CaRoLS deviations: implementation plan

### ISSUE THREAD

**Issue by jalengg** [jalengg/groundwork#1]
# Architectural deviations from CaRoLS that we have not yet fixed

## Context

This repo reproduces (loosely) the CaRoLS paper for road-network synthesis. Through several rounds of audit and experimentation we've closed many gaps with the published architecture, but **one major and several minor deviations remain**. This issue tracks them and proposes an implementation plan.

The most recent successfully-trained model (`diff_safe_fix`, job `8293408`) includes these recently-shipped fixes:

- ✅ **GCI K/V from `fuse(R_c, R_l)` not raw `R_c`** (paper §3.2.2)
- ✅ **LDE `i=3` special case** — at the deepest CDB, skip the concat-with-R_up step (paper §3.2.1)
- ✅ **Conditioning encoder stem** — `Conv 3x3 → BatchNorm → ReLU`, was just `Conv 3x3` (paper §3.2)
- ✅ **Conditioning encoder blocks** — full `Conv(stride=2) → GroupNorm → Swish → Conv 3x3`, was just `Conv(stride=2) → SiLU`

### 1. **CDB topology — additive `R_l + R_up` residual** (high impact, two failed attempts)

Paper §3.2 explicitly says:

> "each CDB comprises a Local Details Enhancement strategy, a Global Context Integration strategy, **and a standard decoder block**: ... the standard decoder block produces an upsampled representation `R_up^i` from `R_g^i`; and the **element-wise addition of `R_l^i` and `R_up^i` is regarded as the CDB's output `R^i`**."

Our current implementation:

- `ConditionAwareDecoderBlock.forward()` returns `GCI(LDE(...), R_c)` — no standard decoder block inside, no `R_l + R_up` additive residual
- The "standard decoder block" lives **outside** the CDB as `UNetBlock dec_i` (with self-attn + residual + upsample + time-embedding)
- The U-Net forward concats CDB output with the noise-stream skip (`e_i`) and feeds to `dec_i`

This is a vanilla U-Net pattern, not the paper's spec.

**Failed attempts to fix:**

| Attempt | What | Failure mode |
|---------|------|---------------|
| v1 (early) | LDE took only `Z_c` (no noise-stream input); StandardDecoderBlock no residual | Loss stuck at ~0.81, salt-and-pepper output, vanishing gradients |
| v2 | v1 + StandardDecoderBlock with residual + self-attn | Loss stuck at ~0.81, same pathology |
| v3 ("full_fix" `8292670`) | Kept paper-spec LDE; CDB output = `R_l + StandardDecoderBlock(R_g)`; removed `dec_i` outside | Loss stuck at ~0.80, grad_norm decay 0.17 → 0.16 |

**Hypotheses for why these fail:**

1. **Time-embedding loss** — removing `dec_i` removes the only path that injects timestep `t_emb` into the decoder output. The std_decoder I wrote doesn't take `t_emb`. The diffusion model loses awareness of which denoising step it's at.
2. **Capacity reduction** — `dec_i` had self-attention, residual, learned upsample (ConvTranspose). Replacing with a smaller std_decoder + interpolation is meaningfully less capable.
3. **Resolution ambiguity in paper** — "R_up^i is upsampled from R_g^i" and "R^i = R_l^i + R_up^i" can't both be true unless R_l and R_up are at the same resolution. Paper text doesn't clarify whether the std_decoder upsamples (in which case R_l must also be upsampled somewhere) or operates at the same resolution as R_g (in which case "upsampled" is misleading).

### 2. **Population density conditioning channel** (medium impact, deliberate omission)

Paper conditioning is `(elev, water, landuse one-hot, **population density**)`. We dropped population density to simplify the data pipeline.

For our use case (Cities Skylines mod with player-painted landuse), population density isn't a natural input. **Likely safe to keep dropped**, but flagging.

### 3. **Connectivity refiner second-stage model** (DRoLaS only, not CaRoLS)

DRoLaS has a second LDM that refines disconnected segments. CaRoLS does not (uses post-processing instead). We have post-processing implemented (`model/postprocess.py`). **Not a CaRoLS gap.**

### 4. **`cfg_prob` (ρ)** (deliberate deviation)

Paper uses `ρ=0.5` (50/50 conditional/unconditional training). We use `ρ=0.1` (90/10) because our use case prioritizes conditional fidelity. Diagnostic in `model/diag_conditioning.py` showed model treats conditioning as a global density knob rather than a spatial guide; lowering `ρ` was an attempt to strengthen spatial conditioning. Effect was modest. **Deliberate, not a bug.**

### 5. **Various small-scale details**

- Noise stream lacks initial `3x3 conv stem` before encoder blocks (paper says "z is processed by a 3×3 convolutional layer **and** 4 encoder blocks")
- `ResBlock` in VAE has extras (GN/SiLU/skip) the paper doesn't strictly specify
- VAE focal loss uses categorical-softmax form, not the literal binary Eq. 2

These are all low-impact and not currently blocking.

## Implementation plan for the CDB topology fix

The CDB topology is the only deviation worth concentrated effort. Approach in increasing order of risk:

### Plan A — Minimal change, preserve `dec_i` capability (low risk)

Don't remove `dec_i`. Add an **explicit additive residual** `R_l → R^i` that runs in parallel to the existing CDB→cat→dec_i path:

```
R_l = LDE(...)
R_g = GCI(R_l, R_c)             # at level i
inner = cat([R_g, e_i], 1)       # (existing)
R_up = dec_i(inner, t_emb)       # (existing, upsamples)
# NEW: project R_l to upsampled resolution and channel count
R_l_proj = upsample_proj(R_l)
return R_up + R_l_proj
```

This adds the additive residual the paper specifies without removing the working `dec_i` capacity. **Test first.**

### Plan B — Restructured StandardDecoderBlock with time-embedding (medium risk)

Rewrite `StandardDecoderBlock` to be functionally equivalent to `UNetBlock` (time-emb, self-attn, residual, learned upsample). Remove `dec_i`. CDB output is `R_l + std_decoder(R_g, t_emb)`.

Critical: `std_decoder` must take `t_emb` to inject diffusion timestep awareness. The two failed attempts were missing this.

### Plan C — Skip the additive residual, keep current architecture (zero risk, accept deviation)

Acknowledge that `R^i = R_l + R_up` is structurally close enough to vanilla U-Net + CDB-as-bottleneck-fusion (which we already do) and don't fight it. Mark it as a permanent deviation in `docs/findings.md`.

## Recommended order

1. Try **Plan A** first. Cheap, isolated change to `unet.py`. If training is healthy, it's done.
2. If Plan A doesn't visibly help residential structure, try **Plan B** with explicit time-emb on the std_decoder.
3. If Plan B also fails, accept **Plan C**.

## Acceptance criteria

A passing implementation should:

1. Train without loss collapse (val loss reaches < 0.25 within 100 epochs on the existing dataset).
2. Match or improve over `diff_safe_fix` on validation loss.
3. Show visible improvement in **spatial conditioning** — running `model/diag_conditioning.py` on the new model should show >40% pixel disagreement between "all residential" and "all commercial" predictions (currently ~30%).
4. Optional: visible improvement in residential parallel-grid texture in the sample images.

## Related artifacts

- `docs/carols_audit_v2.md` — line-by-line CaRoLS-vs-impl audit (generated 2026-04-28)
- `docs/findings.md` — running log of experiments and ruled-out levers
- `model/diag_conditioning.py` — diagnostic script for measuring spatial conditioning use
- `cluster_results/diff_safe_fix/` — current best model with the fixes that DID train (post-this-issue)
